### ISSUE TITLE
Huge performance degradation with discards and inductive datatype generators

### ISSUE THREAD

**Issue by catalin-hritcu** [QuickChick/QuickChick#370]
We were seeing huge performance degradation in one of our developments in the presence of discards for elements of the inductive types representing Imp abstract syntax. I managed to reduce this to a completely self contained example, so the problem seems to be coming from QuickChick itself:
```coq
From QuickChick Require Import QuickChick.

Inductive mynat : Type :=
  | OO
  | S1 (n : mynat)
  | S2 (n : mynat)
  | S3 (n : mynat).

Derive (Arbitrary, Shrink, Show) for mynat.

(* Time Elapsed: 0.100883s  *)
QuickChick (forAll arbitrary (fun (n:mynat) =>
            true)).

(* Time Elapsed: 5.507796s -- with discards *)
QuickChick (forAll arbitrary (fun (n:mynat) =>
            (implication false false))).
```
In our actual development the difference between having and not having discards is many orders of magnitude bigger, but anyway, I hope this simple example already illustrates the problem.

**Comment by catalin-hritcu** [comment]
Actually adding 3 more `S` constructors makes this already completely blow up even if the generated numbers are at most 7(!), so tiny:
```coq
Inductive mynat2 : Type :=
  | OO
  | S1 (n : mynat2)
  | S2 (n : mynat2)
  | S3 (n : mynat2)
  | S4 (n : mynat2)
  | S5 (n : mynat2)
  | S6 (n : mynat2).

Derive (Arbitrary, Shrink, Show) for mynat2.

(* Time Elapsed: 24.980577s *)
QuickChick (forAll arbitrary (fun (n:mynat2) =>
            true)).

(* Time Elapsed: 2824.795694s -- with discards *)
QuickChick (forAll arbitrary (fun (n:mynat2) =>
            (implication false false))).
```
One thing to note is that even in the absence of discards, the ~25s time here is a bit unexpected, but still nothing in comparison to the 2825s this takes with discards.

**Comment by lemonidas** [comment]
This was a much deeper issue - I thought this had been fixed a while ago when @Lysxia first pointed it out way back in 2018, but it must have crept back in. 

The problem is laziness, or rather the lack of it. When a series of generators is combined with `oneof` or `frequency`, it can cause an exponential overhead. Something like `oneof [bind g1 k1; bind g2 k2]` looks innocuous, but `g1` and `g2` are evaluated eagerly as arguments to the bind (which includes recursive calls). For now, [this branch](https://github.com/QuickChick/QuickChick/tree/thunk-in-derivations) fixes the problem by deriving correct (thunked) versions of generators for both Arbitrary and ArbitraryST. It fails some tests that I'll need to update, but it should be of immediate help.

In general though, we'll need to think how to stop handing users a very easy way of shooting themselves in the foot in the future.

**Comment by catalin-hritcu** [comment]
Thanks a lot @lemonidas. I see that your PR fixing this for derived generators (#373) got merged yesterday.

What to do about hand-written generators though? Does one need to manually add `thunkGen` for each element of the list whenever using `oneof` or `frequency` for defining recursive generators? And if so, should this be documented everywhere? :)

As you seem to imply above though,  a solution that makes it harder for people to shoot themselves in the foot would be much better. Change these operators to always take thunks as already proposed in #73? :)

Noticed this late, but anyway here are some related issues:
- #229 which I exactly duplicated here, sorry
- #73 and maybe even #20 from 9 years ago? :)

**Comment by lemonidas** [comment]
Yeah, the "proper" solution would probably be to make the combinators take thunks, and provide useful notation (which we already do to hide the default element). I'll give it a shot today.
